### DISCUSSION TITLE
Improving asynchronous operations by adding ToDictionaryAsync

### DISCUSSION THREAD

**Discussion by ivanlegin** [ravendb/ravendb#19605] [Ideas]
Hello, is there a chance to implement a `ToDictionaryAsync()` method? 

While it’s not difficult to construct it from a list or array, instantiating those collections solely to serve as a source for a dictionary feels unnecessary and impractical. 

For example:  
`(await query.ToListAsync()).ToDictionary(i => i.Id);`

Especially since in the case of a non-asynchronous call, this works without requiring an intermediate collection by design:  
`query.ToDictionary(i => i.Id);`

What is your opinion on this?  

Also, `LazilyAsync()` always returns a list (despite the method signature stating `IEnumerable`). If `ToDictionaryAsync()` is implemented, it might be worth considering offering the user a choice of the collection type they want.

**Comment by karmeli87** [comment]
Hi, 

I've open a ticket to add this functionality https://issues.hibernatingrhinos.com/issue/RavenDB-23179

Not sure if or when it is going to be scheduled.

If you're interested, we'd love to see a pull request from you. Feel free to contribute. :-)
